x/gov/types/msgs.go

```diff
@@ -32,64 +39,107 @@ type MsgSubmitProposalI interface {
 	SetProposer(sdk.AccAddress)
 }
 
-// NewMsgSubmitProposalBase creates a new MsgSubmitProposalBase.
-func NewMsgSubmitProposalBase(initialDeposit sdk.Coins, proposer sdk.AccAddress) MsgSubmitProposalBase {
-	return MsgSubmitProposalBase{
+// NewMsgSubmitProposal creates a new MsgSubmitProposal.
+func NewMsgSubmitProposal(content Content, initialDeposit sdk.Coins, proposer sdk.AccAddress) (*MsgSubmitProposal, error) {
+	m := &MsgSubmitProposal{
 		InitialDeposit: initialDeposit,
 		Proposer:       proposer,
 	}
+	err := m.SetContent(content)
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
 }
 
-func (msg *MsgSubmitProposalBase) GetInitialDeposit() sdk.Coins { return msg.InitialDeposit }
+func (m *MsgSubmitProposal) GetInitialDeposit() sdk.Coins { return m.InitialDeposit }
 
-func (msg *MsgSubmitProposalBase) GetProposer() sdk.AccAddress { return msg.Proposer }
+func (m *MsgSubmitProposal) GetProposer() sdk.AccAddress { return m.Proposer }
+
+func (m *MsgSubmitProposal) GetContent() Content {
+	content, ok := m.Content.GetCachedValue().(Content)
+	if !ok {
+		return nil
```

Comment:
Req: The CachedValue can be nil (in tests for example). It would be consistent to unpack the Any type. 
Q: I am wondering about the interface cast: When `m.Content` is nil then nil makes sense as a result but when "CachedValue" does not implement the `Content` interface why returning nil instead of an error. This case should not happen with proper use of SetContent but `m.Content` is public so can be anything.

Reply:
Is there a better approach you would suggest?

Reply:
🤔 Ah, you don't have access to the unpacker without changing the method signature....
I guess you don't want to add it as a parameter but rely on the stack to initialize the object proper. As a safety net I would add the following conditions so that the method can not return anything inconsistent.
```go
func (m *MsgSubmitProposal) GetContent() Content {
	if m.Content == nil {
		return nil
	}
	cv := m.Content.GetCachedValue()
	if cv == nil {
		panic("content was not unpacked")
	}
	content, ok := cv.(Content)
	if !ok {
		panic(fmt.Sprintf("illegal type: %T", cv))
	}
	return content
}
```